raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+        """
+        Evaluates the interpolating function.
+
+        :param double x: the point for which an interpolated value is required
+        :return: the interpolated value at point x.
+        """
+        cdef int index = find_index(self._x, x)
+
+        if index == -1:
+            if x < self._x[0] - self._extrapolation_range:
+                raise ValueError(
+                    f'The specified value (x={x}) is outside of extrapolation range.')
+            return self._extrapolator.evaluate(x, index)
+        elif index == self._last_index:
+            if x > self._x[self._last_index] + self._extrapolation_range:
+                raise ValueError(
+                    f'The specified value (x={x}) is outside of extrapolation range.')
+            return self._extrapolator.evaluate(x, index)
+        else:
+            return self._interpolator.evaluate(x, index)
+
+    @property
+    def domain(self):
+        """
+        Returns bounding box of the provided inputs.
+        Order: min(x), max(x), min(f), max(f)
+        :warning: doesn't take extrapolator into account at the moment
+        """
+        return np.min(self._x), np.max(self._x), np.min(self._f), np.max(self._f)
+
+
+cdef class _Interpolator1D:
+    """Base class for 1D interpolators. """
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        """
+        Calculates interpolated value at given point. 
+    
+        :param double px: the point for which an interpolated value is required
+        :param int index: the lower index of the bin containing point px. (Result of bisection search).   
+        """
+        raise NotImplementedError('_Interpolator is an abstract base class.')
+
+
+cdef class _Interpolator1DLinear(_Interpolator1D):
+    """
+    Linear interpolation of 1D function.
+
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param Extrapolator1D extrapolator: extrapolator object
+    """
+    def __init__(self, double[::1] x, double[::1] f):
+        self._x = x
+        self._f = f
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        return linear1d(self._x[index], self._x[index + 1], self._f[index], self._f[index + 1], px)
+
+
+cdef class _Interpolator1DCubic(_Interpolator1D):
+    """
+    Cubic interpolation of 1D function
+
+    When called, stores cubic polynomial coefficients from the value of the function at the neighboring spline points
+    and the gradient at the neighbouring spline points based on central difference gradients. The polynomial
+    coefficients and gradients are calculated between each spline knots normalised to between 0 and 1.
+
+    :param x: 1D memory view of the spline point x positions.
+    :param f: 1D memory view of the function value at spline point x positions.
+    """
+    def __init__(self, double[::1] x, double[::1] f):
+        self._x = x
```

Comment:
Storing the memory views is common to all the sub-classes, move to the base class.